modules/fx/arrow-fx/src/main/kotlin/arrow/fx/Queue.kt

```diff
@@ -0,0 +1,231 @@
+   * If the queue is already shutdown, the `IO` will resume right away.
+   */
+  fun awaitShutdown(): Kind<F, Unit> =
+    Promise<F, Unit>(CF).flatMap { promise ->
+      val io = promise.complete(Unit)
+      ref.modify { state ->
+        state.fold(
+          ifSurplus = { it.copy(shutdownHook = it.shutdownHook.followedBy(io.unit())) toT unit() },
+          ifDeficit = { it.copy(shutdownHook = it.shutdownHook.followedBy(io.unit())) toT unit() },
+          ifShutdown = { state toT io.unit() }
+        )
+      }.flatten().followedBy(promise.get())
+    }
+
+  /**
+   * Cancels any fibers that are suspended on `offer` or `take`.
+   * Future calls to `offer*` and `take*` will be interrupted immediately.
+   */
+  fun shutdown(): Kind<F, Unit> = ref.modify { state ->
+    state.fold(
+      ifSurplus = {
+        if (it.putters.isEmpty()) State.Shutdown(CF) toT it.shutdownHook
+        else {
+          val forked = forkAll(it.putters.toList()
+            .map { (_, p) -> p.error(QueueShutdown) }).flatMap { it.join() }
+          State.Shutdown(CF) toT (forked.followedBy(it.shutdownHook))
+        }
+      },
+      ifDeficit = {
+        if (it.takers.isEmpty()) State.Shutdown(CF) toT it.shutdownHook
+        else {
+          val forked = forkAll(it.takers.toList()
+            .map { it.error(QueueShutdown) }).flatMap { it.join() }
+          State.Shutdown(CF) toT (forked.followedBy(it.shutdownHook))
+        }
+      },
+      ifShutdown = { state toT unit() }
+    )
+  }.flatten()
+
+  companion object {
+    fun <F, A> bounded(capacity: Int, CF: Concurrent<F>): Kind<F, Queue<F, A>> = CF.run {
+      Ref<State<F, A>>(State.Surplus(IQueue.empty(), IQueue.empty(), this, unit())).map {
+        Queue(capacity, it, this)
+      }
+    }
+
+    internal fun <F, A, C> State<F, A>.fold(
+      ifSurplus: (State.Surplus<F, A>) -> C,
+      ifDeficit: (State.Deficit<F, A>) -> C,
+      ifShutdown: (State.Shutdown<F>) -> C
+    ) =
+      when (this) {
+        is State.Surplus -> ifSurplus(this)
+        is State.Deficit -> ifDeficit(this)
+        is State.Shutdown -> ifShutdown(this)
+      }
+  }
+
+  private fun <A> removeTaker(taker: Promise<F, A>): Kind<F, Unit> =
+    ref.update { state ->
+      state.fold(
+        ifSurplus = ::identity,
+        ifDeficit = { it.run { copy(takers.filterNot { t -> t == taker }) } },
+        ifShutdown = ::identity
+      )
+    }
+
+  private fun removePutter(putter: Promise<F, Unit>): Kind<F, Unit> =
+    ref.update { state ->
+      state.fold(
+        ifSurplus = { it.run { copy(putters = putters.filterNot { p -> p == putter }) } },
+        ifDeficit = ::identity,
+        ifShutdown = ::identity
+      )
+    }
+}
+
+object QueueShutdown : RuntimeException() {
+  override fun fillInStackTrace(): Throwable = this
+}
+
+// Write fiber Concurrent typeclass??
+fun <F, A, B, C> Fiber<F, A>.zipwith(fb: Fiber<F, B>, M: Monad<F>, f: (Tuple2<A, B>) -> C): Fiber<F, C> {
```

Comment:
This PR will add `Monoid` & `Applicative` instance for `Fiber` which are based on `Concurrent` so that should solve this problem.

https://github.com/arrow-kt/arrow/pull/1783